src/wallet/bdb.cpp

```diff
@@ -363,6 +364,13 @@ void BerkeleyDatabase::Open()
 
             m_db.reset(pdb_temp.release());
 
+            auto batch = MakeBatch();
+            if (!batch->Exists(std::string("walletid"))) {
```

Comment:
I think this is in the wrong place. From a quick look `CWallet::LoadWallet` seems a better place, like:

```cpp
    if (wallet_id.IsNull()) {
        if (database->Format() == "dbd") {
             // use BDB ID
        } else {
             // generate random id
        }
        // persist
    }
    return DBErrors::LOAD_OK;
}
```
This would allow you to drop `EnsureWalletIDWithDB`. Only bad thing here is to compare agains "bdb". Would be better to add  `Format()` and `FormatString()`.

Reply:
I've moved all of the logic into `EnsureWalletIDWithDB`.